packages/vite/src/node/plugins/asset.ts

```diff
@@ -353,11 +355,48 @@ async function fileToBuiltUrl(
     const { search, hash } = parseUrl(id)
     const postfix = (search || '') + (hash || '')
     const output = config.build?.rollupOptions?.output
-    const assetFileNames =
+
+    // Steps to determine which assetFileNames will be actually used.
+    const defaultAssetFileNames = path.posix.join(
+      config.build.assetsDir,
+      '[name].[hash][extname]'
+    )
+    // First, if output is an object, use assetFileNames in it.
+    // And a default assetFileNames as fallback.
+    let assetFileNames: Exclude<OutputOptions['assetFileNames'], undefined> =
       (output && !Array.isArray(output) ? output.assetFileNames : undefined) ??
-      // defaults to '<assetsDir>/[name].[hash][extname]'
-      // slightly different from rollup's one ('assets/[name]-[hash][extname]')
-      path.posix.join(config.build.assetsDir, '[name].[hash][extname]')
+      defaultAssetFileNames
+    if (output && Array.isArray(output)) {
+      // Second, if output is an array, adopt assetFileNames in the first object.
+      assetFileNames = output[0].assetFileNames ?? assetFileNames
+      // Third, check if assetFileNames among output array return the same result.
+      // If not, log a warn and keep using the first assetFileNames.
+      const filenameSet = output.reduce(
+        (cumulate, { assetFileNames = defaultAssetFileNames }) => {
+          const filename =
+            typeof assetFileNames === 'string'
+              ? assetFileNames
+              : assetFileNames({
+                  type: 'asset',
+                  name: file,
+                  source: content
+                })
+          cumulate.add(filename)
+          return cumulate
+        },
+        new Set<string>()
+      )
+      if (filenameSet.size > 1) {
```

Comment:
I think that we can ask users to use the same `assetFileNames` reference if they want to use the function form. This would allow us to move the check to `config.ts` while resolving the config so we don't need to do it every time assetFileNames is used. The check could also be done with something like `output.some((o) => o && o.assetFileNames !== assetFileNames)` in this case.

Reply:
I agree with your opinion. Checks for every asset is redundant. 
But function reference may not be an optimal way. For example, some plugins may wrap user options, and assetFilesNames is replaced with a new function. Or people may just want to emit asset files to different dir(this is why output can be an array, i guess).

Perhaps we can invoke assetFileNames functions only once in config.ts, passing a virtual file to do a quick check. So that redundant checks and user being forced to use same assetFileNames function are both avoided.

Reply:
The problem is that the quick check wouldn't give you enough information, since as you say, a function could point to different folders and differ for only a subset of files.
Do you have an example of a plugin replacing `assetFileNames` with a new function? I think that this may not be an issue. I would first try to do it in this way, and only add complexity if we find we are breaking real use cases.

Reply:
Legacy-plugin [does modification to entryFileNames and chunkFileNames](https://github.com/vitejs/vite/blob/2dfc0152be6ed531fdebe7dc76314ac9f80b7b1d/packages/plugin-legacy/src/index.ts#L225). That's why I'm worried about reference check. There may comes the same with assetFileNames in the future.
Anyway, I'll move the check to config.ts and check via function reference.

Reply:
You're right about being worried here. Let's try it out with the global check first, and keep an eye on this. Another option is to cache the check in case there are no functions or the ref check is ok, and fallback to your current approach if that isn't the case. But let's get that real use case first before going there 👍🏼